pennylane/hf/observable.py

```diff
@@ -0,0 +1,327 @@
+        op = [((op[0], 1), (op[1], 1), (op[2], 0), (op[3], 0))]
+
+        if op[0][0][0] == op[0][1][0] or op[0][2][0] == op[0][3][0]:
+            return 0
+
+    for t in op:
+        for l in t:
+            z = [(index, "Z") for index in range(l[0])]
+            x = z + [(l[0], "X"), 0.5]
+
+            if l[1]:
+                y = z + [(l[0], "Y"), -0.5j]
+
+            else:
+                y = z + [(l[0], "Y"), 0.5j]
+
+            if t.index(l) == 0:
+                q = [x, y]
+            else:
+                m = []
+                for t1 in q:
+                    for t2 in [x, y]:
+                        q1, c1 = _pauli_mult(t1[:-1], t2[:-1])
+                        m.append(q1 + [c1 * t1[-1] * t2[-1]])
+                q = m
+
+    c = [p[-1] for p in q]
+    o = [p[:-1] for p in q]
+
+    for item in o:
+        k = [i for i, x in enumerate(o) if x == item]
+        if len(k) >= 2:
+            for j in k[::-1][:-1]:
+                del o[j]
+                c[k[0]] = c[k[0]] + c[j]
+                del c[j]
+
+    pauli_map = {"X": qml.PauliX, "Y": qml.PauliY, "Z": qml.PauliZ}
+    for i, term in enumerate(o):
+        if len(term) == 0:
+            o[i] = qml.Identity(0)
+        if len(term) == 1:
+            o[i] = pauli_map[term[0][1]](term[0][0])
+        if len(term) > 1:
+            k = [pauli_map[t[1]](t[0]) for t in term]
+            o[i] = reduce(lambda x, y: x @ y, k)
+
+    return c, o
+
+
+def simplify(h, cutoff=1.0e-12):
+    r"""Add together identical terms in the Hamiltonian.
+
+    The Hamiltonian terms with identical Pauli words are added together and eliminated if the
+    overall coefficient is smaller than a cutoff value.
+
+    Args:
+        h (Hamiltonian): PennyLane Hamiltonian
+        cutoff (float): cutoff value for discarding the negligible terms
+
+    Returns:
+        Hamiltonian: Simplified PennyLane Hamiltonian
+
+    **Example**
+
+    >>> c = np.array([0.5, 0.5])
+    >>> h = qml.Hamiltonian(c, [qml.PauliX(0) @ qml.PauliY(1), qml.PauliX(0) @ qml.PauliY(1)])
+    >>> print(simplify(h))
+    (1.0) [X0 Y1]
+    """
+    wiremap = dict(zip(h.wires, range(len(h.wires) + 1)))
+
+    c = []
+    o = []
+    for i, op in enumerate(h.ops):
+        op = qml.operation.Tensor(op).prune()
+        op = qml.grouping.pauli_word_to_string(op, wire_map=wiremap)
+        if op not in o:
+            c.append(h.coeffs[i])
+            o.append(op)
+        else:
+            c[o.index(op)] += h.coeffs[i]
+
+    coeffs = []
+    ops = []
+    nonzero_ind = np.argwhere(abs(np.array(c)) > cutoff).flatten()
+    for i in nonzero_ind:
+        coeffs.append(c[i])
+        ops.append(qml.grouping.string_to_pauli_word(o[i], wire_map=wiremap))
+
+    try:
+        coeffs = qml.math.stack(coeffs)
+    except ValueError:
+        pass
+
+    return qml.Hamiltonian(coeffs, ops)
+
+
+def _pauli_mult(p1, p2):
+    r"""Return the result of multiplication between two tensor products of Pauli operators.
```

Comment:
@soranjh this one shouldn't be a private method here, but a public method in `qml.grouping`. It is _super_ useful, and I know other parts of the codebase will need to use it!

If we leave it here private, no-one will know it is available.

Reply:
In fact there are already two very similar functions there:

- [`qml.grouping.pauli_mult`](https://pennylane.readthedocs.io/en/latest/code/api/pennylane.grouping.pauli_mult.html)
- [`qml.grouping.pauli_mult_with_phase`](https://pennylane.readthedocs.io/en/latest/code/api/pennylane.grouping.pauli_mult_with_phase.html)

Reply:
Hi @josh146. This `jordan_wigner` function uses its own notation for Pauli terms and this `_pauli_mult` function allows for an efficient multiplication under that notation. The reason for this different notation is _only_ efficiency. I have another version of the `jordan_wigner` function which uses PL operator arithmetic but it is not very efficient. When we support efficient operator/observable arithmetic in PL, there will be no need for this version of `_pauli_mult` and we can simply delete it.

Reply:
Thanks for the clarification @soranjh. I agree with your assessment, let's keep this here for now as efficiency is so important to build Hamiltonians

Reply:
thanks for clarifying @soranjh, makes sense!